### PR TITLE
Recurring drives frontend

### PR DESCRIPTION
**Pull request by fynsta**
Redoing #135 which I accidentally merged.

We're truly sorry, this is the biggest feature by far. This is the front end part of the recurring drives!
Instead of all the new changes, I'm going to describe a flow for testing:

You can create recurring drives by clicking the checkbox on the regular create drive page. Please check if everything is understandable for usability.
You are taken to the new recurring drive detail page, where you can see your schedule, information and all generated drives. In addition to that, you can edit the schedule by tapping the icon on the top right.
You can see your recurring drives on the new "Recurring" tab under the Drives navigator bar entry.
The rest of the time was essentially spent on figuring out edge cases and weirdness that has to do with the complexity of situations that can arise by cancelling, changing or picking weird recurring drive schedules.

### CODE REVIEW

**Comment by fynsta** [comment]
Open feedback by @istzen:

> i have an overflow error, at the elements that show on which days the drive will be.

> when i stop the recurring drive, the button should change so that it's clear that it is really stopped. And if i stop a drive and then change the rule, i cannot stop it anymore, the drives are still there.

**Comment by istzen** [comment]
image

File: lib/drives/util/recurrence_options_edit.dart [hunk 1]
```diff
@@ -233,10 +234,13 @@ class RecurrenceOptionsEditState extends State<RecurrenceOptionsEdit> {
   }
 
   Widget buildIndicator() {
-    return RecurrenceOptionsIndicator(
-      before: originalRecurrenceRule,
-      after: recurrenceOptions.recurrenceRule,
-      start: recurrenceOptions.startedAt.isAfter(DateTime.now()) ? recurrenceOptions.startedAt : DateTime.now(),
+    return ExpandableSection(
```
Comment (fynsta) [inline lib/drives/util/recurrence_options_edit.dart:237]: This still needs to set the parameter to decide whether it is expanded by default

File: lib/l10n/app_de.arb [hunk 2]
```diff
@@ -16,6 +16,7 @@
   "remove": "Entfernen",
   "before": "Vorher",
   "after": "Nachher",
+  "preview": "Vorschau",
```
Comment (fynsta) [inline lib/l10n/app_de.arb:19]: Maybe "Vorschau anzeigen" would be better?

File: lib/drives/util/recurrence_options_edit.dart (hunk 1 above)
Comment (Eddie-42) [reply lib/drives/util/recurrence_options_edit.dart:237]: Yes I know. Working on that tomorrow

File: lib/l10n/app_de.arb (hunk 2 above)
Comment (Eddie-42) [reply lib/l10n/app_de.arb:19]: "Vorschau anzeigen" if it's not expanded, "Vorschau" if it's expanded?

Comment (fynsta) [reply lib/l10n/app_de.arb:19]: That seems weird... Why not "Vorschau verstecken"?

Comment (Eddie-42) [reply lib/l10n/app_de.arb:19]: You know there is a "Ausklappen" or "Einklappen" button next to that right?

Comment (fynsta) [reply lib/l10n/app_de.arb:19]: Maybe I should have tested this out :D Then nevermind

File: lib/drives/util/recurrence_options_indicator.dart [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      ? Theme.of(context).own().onSuccess
+      : type == RecurrenceOptionsIndicatorType.removed
+          ? Theme.of(context).colorScheme.onError
+          : Theme.of(context).colorScheme.onPrimary;
+
+  @override
+  Widget build(BuildContext context) {
+    final TextStyle dateStyle = TextStyle(color: getForegroundColor(context));
+    final TextPainter textPainter = TextPainter(
+      text: TextSpan(text: localeManager.formatDate(DateTime(2022, 12, 12)), style: dateStyle),
```
Comment (fynsta) [inline lib/drives/util/recurrence_options_indicator.dart:88]: funny... Maybe add a comment here to make sure people understand what this is doing

Also, maybe there are languages where this is not the longest, but I don't care...

File: lib/drives/util/recurrence.dart [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+      : occurrences! >= 100
+          ? S.of(context).recurrenceEndValidationOccurrencesTooMany
+          : null;
+
+  @override
+  bool operator ==(Object other) =>
+      other is RecurrenceEndChoiceOccurrence && isCustom == other.isCustom && occurrences == other.occurrences;
+
+  @override
+  RecurrenceEndChoice copyWith({bool? isCustom}) =>
```
Comment (fynsta) [inline lib/drives/util/recurrence.dart:211]: This is not covered, maybe it's not even needed anymore?

File: lib/util/extensions/time_of_day_extension.dart [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+extension TimeOfDayExtension on TimeOfDay {
+  bool isBefore(TimeOfDay other) {
+    return hour < other.hour || (hour == other.hour && minute < other.minute);
+  }
+
+  bool isAfter(TimeOfDay other) {
+    return hour > other.hour || (hour == other.hour && minute > other.minute);
+  }
+
+  bool isAtSameMomentAs(TimeOfDay other) {
```
Comment (fynsta) [inline lib/util/extensions/time_of_day_extension.dart:12]: Also maybe not needed?

File: lib/drives/util/recurrence.dart (hunk 4 above)
Comment (Eddie-42) [reply lib/drives/util/recurrence.dart:211]: Yeah I saw that... Wasn't that the one where some arguments were never called? I thought I'd leave that for consistency. But there are some places in the code where I think some parameters/methods are not needed, lmk if you think I should just delete those

**Review by Eddie-42 — APPROVED** [review]

File: lib/drives/pages/recurring_drive_edit_page.dart [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    widgets.add(TripOverview(_recurringDrive));
+    widgets.add(const Divider(thickness: 1));
+
+    widgets.add(
+      Form(
+        key: _formKey,
+        child: RecurrenceOptionsEdit(
+          recurrenceOptions: recurrenceOptions,
+          predefinedEndChoices: const <RecurrenceEndChoice>[],
+          showPreview: _defaultPreviewExpanded ?? true,
```
[the remark this answers is not in the document]
Comment (fynsta) [reply lib/drives/pages/recurring_drive_edit_page.dart:79]: das versteht man nicht, so viel haben wir in der Nutzerstudie gesehen

[the remark this answers is not in the document]
Comment (Showcas) [reply lib/drives/pages/recurring_drive_edit_page.dart:79]: mhh ok, weiß hlat nicht weil der button ein wenig out of place vorkommt aber an sonsten ist der rest ok

**Review by Showcas — APPROVED** [review]

**Pull request merged**